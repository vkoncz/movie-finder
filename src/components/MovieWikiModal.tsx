import {
  IonButton,
  IonCol,
  IonContent,
  IonIcon,
  IonRow,
  IonSkeletonText,
  IonTitle,
  IonToolbar,
} from '@ionic/react';
import React, { useEffect, useState } from 'react';
import { wikiSummaryClient } from '../data/fetch/wikiSummaryClient';
import { chevronBack, openOutline } from 'ionicons/icons';
import { MovieInfo } from '../models/MovieInfo';

interface Props {
  movieInfo: MovieInfo;
  modalClose: () => void;
}

export const MovieWikiModal: React.FC<Props> = ({ movieInfo, modalClose }) => {
  const [wikiSummary, setWikiSummary] = useState('');
  const [wikiLink, setWikiLink] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [missingWiki, setMissingWiki] = useState(false);

  useEffect(() => {
    setIsLoading(true);
    setMissingWiki(false);

    wikiSummaryClient(movieInfo.title)
      .then(result => {
        setWikiSummary(result.extract);
        setWikiLink(result.content_urls.desktop.page);
        setIsLoading(false);
      })
      .catch(() => {
        setIsLoading(false);
        setMissingWiki(true);
      });
  }, [movieInfo]);

  return (
    <>
      <IonToolbar color="medium">
        <IonTitle>{movieInfo.title}</IonTitle>
      </IonToolbar>
      <IonContent>
        <p className="ion-padding-start ion-padding-end">
          {isLoading ? (
            <>
              <IonSkeletonText animated style={{ width: '60%' }} />
              <IonSkeletonText animated />
              <IonSkeletonText animated style={{ width: '90%' }} />
              <IonSkeletonText animated style={{ width: '70%' }} />
            </>
          ) : (
            wikiSummary
          )}
          {missingWiki ? 'Cannot found Wikipedia article' : ''}
        </p>

        {!isLoading && !missingWiki ? (
          <IonRow>
            <IonButton href={wikiLink} target="_blank" color="light">
              Open Wikipedia article
              <IonIcon icon={openOutline} />
            </IonButton>
          </IonRow>
        ) : (
          ''
        )}

        <IonRow>
          <IonButton
            href={`https://www.imdb.com/title/${movieInfo.imdb}`}
            target="_blank"
            color="light"
          >
            IMDB
            <IonIcon icon={openOutline} />
          </IonButton>
        </IonRow>

        <IonRow>
          <IonCol class="ion-text-center">
            <IonButton
              color="warning"
              class="ion-align-self-center"
              shape="round"
              onClick={modalClose}
            >
              <IonIcon icon={chevronBack} />
            </IonButton>
          </IonCol>
        </IonRow>
      </IonContent>
    </>
  );
};
