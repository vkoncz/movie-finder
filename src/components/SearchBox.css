ion-button {
  height: 42px;
  width: 60px;
}

ion-grid {
  max-width: 1000px;
}
