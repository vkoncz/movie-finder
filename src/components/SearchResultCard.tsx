import {
  IonCard,
  IonCardContent,
  IonCardHeader,
  IonCardTitle,
  IonIcon,
  IonItem,
  IonLabel,
} from '@ionic/react';
import React from 'react';
import { Details } from '../models/MovieDetails';
import { star, pricetag } from 'ionicons/icons';

interface Props {
  details: Details;
  onMovieClick: (title: string) => void;
}

export const SearchResultCard: React.FC<Props> = ({ details, onMovieClick }) => {
  return (
    <IonCard
      button={true}
      onClick={() => {
        onMovieClick(details.title);
      }}
    >
      <IonCardHeader color="medium">
        <IonCardTitle>{details.title}</IonCardTitle>
      </IonCardHeader>

      <IonCardContent>
        <IonItem lines="none">
          <IonIcon icon={star} slot="start" size="small"></IonIcon>
          <IonLabel>{details.rating}</IonLabel>
        </IonItem>
        <IonItem lines="none">
          <IonIcon icon={pricetag} slot="start" size="small"></IonIcon>
          <IonLabel>{details.genres.map(genre => genre.name).join(', ')}</IonLabel>
        </IonItem>
      </IonCardContent>
    </IonCard>
  );
};
