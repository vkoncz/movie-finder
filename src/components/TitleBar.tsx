import React from 'react';
import { IonTitle, IonToolbar } from '@ionic/react';

export const TitleBar: React.FC = () => {
  return (
    <IonToolbar color="primary">
      <IonTitle>Movie Finder</IonTitle>
    </IonToolbar>
  );
};
