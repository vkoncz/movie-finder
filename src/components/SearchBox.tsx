import React, { useState } from 'react';
import {
  IonButton,
  IonCol,
  IonRow,
  IonGrid,
  IonIcon,
  IonSearchbar,
  IonToast,
  IonLoading,
} from '@ionic/react';
import { search, videocam } from 'ionicons/icons';
import './SearchBox.css';
import { Details } from '../models/MovieDetails';
import { graphSearchClient } from '../data/fetch/graphSearchClient';

interface Props {
  resultChange: (results?: Details[]) => void;
}

export const SearchBox: React.FC<Props> = ({ resultChange }) => {
  const [searchTerm, setSearchTerm] = useState('');
  const [showError, setShowError] = useState(false);
  const [isLoading, setIsLoading] = useState(false);

  const fetchData = () => {
    setIsLoading(true);
    graphSearchClient(searchTerm)
      .then(result => {
        resultChange(result);
        setIsLoading(false);
      })
      .catch(error => {
        console.log('Connection error:', error);
        setShowError(true);
        setIsLoading(false);
      });
  };

  return (
    <>
      <IonGrid class="ion-center search-container">
        <IonRow class="ion-align-items-center">
          <IonCol class="ion-no-padding">
            <IonSearchbar
              autocomplete="on"
              placeholder="Search movies"
              searchIcon={videocam}
              value={searchTerm}
              onIonChange={e => setSearchTerm(e.detail.value!)}
              onKeyUp={e => {
                if (e.key === 'Enter') fetchData();
              }}
            />
          </IonCol>
          <IonCol size="auto" class="ion-no-padding ion-align-self-center">
            <IonButton onClick={fetchData} class="search-button">
              <IonIcon icon={search} />
            </IonButton>
          </IonCol>
        </IonRow>
      </IonGrid>

      <IonLoading isOpen={isLoading} message="Searching movies..." />

      <IonToast
        isOpen={showError}
        onDidDismiss={() => setShowError(false)}
        message="Connection error, please try it again later"
        duration={5000}
      />
    </>
  );
};
