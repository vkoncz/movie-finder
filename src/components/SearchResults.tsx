import { IonCol, IonGrid, IonModal, IonRow, IonText } from '@ionic/react';
import React, { useState } from 'react';
import { Details } from '../models/MovieDetails';
import { MovieInfo } from '../models/MovieInfo';
import { MovieWikiModal } from './MovieWikiModal';
import { SearchResultCard } from './SearchResultCard';

interface Props {
  results?: Details[];
}

export const SearchResults: React.FC<Props> = ({ results }) => {
  const message = !results || results?.length ? '' : 'No results found';
  const [selectedMovieInfo, setSelectedMovieInfo] = useState<MovieInfo>({ imdb: '', title: '' });
  const [modalVisible, setModalVisible] = useState(false);

  return (
    <>
      <IonGrid>
        <IonRow>
          {results
            ? results.map((result, key) => (
                <IonCol key={key} sizeXl="3" sizeLg="4" sizeMd="6" size="12">
                  <SearchResultCard
                    details={result}
                    onMovieClick={title => {
                      setModalVisible(true);
                      setSelectedMovieInfo(title);
                    }}
                  />
                </IonCol>
              ))
            : ''}
          <IonText>{message}</IonText>
        </IonRow>
      </IonGrid>

      <IonModal isOpen={modalVisible} onDidDismiss={() => setModalVisible(false)}>
        <MovieWikiModal movieInfo={selectedMovieInfo} modalClose={() => setModalVisible(false)} />
      </IonModal>
    </>
  );
};
