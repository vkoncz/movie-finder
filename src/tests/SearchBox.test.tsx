import React from 'react';
import { render } from '@testing-library/react';
import { SearchBox } from '../components/SearchBox';

it('should match to SearchBox snapshots', () => {
  const { asFragment } = render(<SearchBox resultChange={() => {}} />);

  expect(asFragment()).toMatchSnapshot();
});
