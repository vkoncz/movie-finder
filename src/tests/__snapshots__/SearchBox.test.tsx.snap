// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match to SearchBox snapshots 1`] = `
<DocumentFragment>
  <ion-grid
    class="ion-center search-container"
  >
    <ion-row
      class="ion-align-items-center"
    >
      <ion-col
        class="ion-no-padding"
      >
        <ion-searchbar
          autocomplete="on"
          placeholder="Search movies"
          search-icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Videocam</title><path d='M464 384.39a32 32 0 01-13-2.77 15.77 15.77 0 01-2.71-1.54l-82.71-58.22A32 32 0 01352 295.7v-79.4a32 32 0 0113.58-26.16l82.71-58.22a15.77 15.77 0 012.71-1.54 32 32 0 0145 29.24v192.76a32 32 0 01-32 32zM268 400H84a68.07 68.07 0 01-68-68V180a68.07 68.07 0 0168-68h184.48A67.6 67.6 0 01336 179.52V332a68.07 68.07 0 01-68 68z'/></svg>"
          value=""
        />
      </ion-col>
      <ion-col
        class="ion-no-padding ion-align-self-center"
        size="auto"
      >
        <ion-button
          class="search-button"
        >
          <ion-icon
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Search</title><path d='M456.69 421.39L362.6 327.3a173.81 173.81 0 0034.84-104.58C397.44 126.38 319.06 48 222.72 48S48 126.38 48 222.72s78.38 174.72 174.72 174.72A173.81 173.81 0 00327.3 362.6l94.09 94.09a25 25 0 0035.3-35.3zM97.92 222.72a124.8 124.8 0 11124.8 124.8 124.95 124.95 0 01-124.8-124.8z'/></svg>"
          />
        </ion-button>
      </ion-col>
    </ion-row>
  </ion-grid>
</DocumentFragment>
`;
